### PR TITLE
Substantial rewrite to support ND-LAr reco

### PR DESCRIPTION
**Pull request by chenel**
This pull request would introduce a fairly large rewrite to `ND_CAFMaker` in order to support filling of ND-LAr reco branches into output CAFs.  What changed:

- Reorganize the files so there's less clutter at top level
- Introduce a FHICL-based configuration system so that parameters are more easily versioned and modified (instead of the command-line-switch-based system)
- Move the pseudo-reco code out of the top-level `makeCAF.cxx` (see also below)
- Support for / dependency on more "structured" `StandardRecord` from [`duneanaobj`](https://github.com/DUNE/duneanaobj)  ~requires its [PR #1](https://github.com/DUNE/duneanaobj/pull/1) to be merged first~ ✔️ 
- Introduce a base class `IRecoBranchFiller` and the skeleton for switching which concrete chain of reco objects actually gets filled (whether ND-LAr only, ND-LAr + TMS, ND-LAr + ND-GAr, ...)
- Rearrange the "parameterized reco" code as a concrete `IRecoBranchFiller`, and leave it on by default (used unless an mlreco input file is passed)
- Implement a concrete `IRecoBranchFiller` that reads the HD5 `.h5` summary file outputs from the ND-LAr reco and fills the reco branches with them

**CAVEATS**
- Adding the xsec weights to the new `StandardRecord` appears not to be fully working.  Many of them (all those directly copied from GENIE knobs?) seem to be basically ok, but some (I think only those directly implemented in `nusystematics`?) aren't producing the same values.  (I have a Python notebook that compares an old CAF and a new CAF made from the same GENIE input which can be shared.)  **After discussions with Luke P., the interaction model weights have been disabled until the DIRT-II group finishes the model & uncertainty work.**
- Relatedly (?), the CAFAna code that automatically calculates a default central value weight (`FixupRecord()`) needs to be disabled to use these CAFs.  I have a branch of `lblpwgtools` [`feature/ndlar-cafs-wip`](https://github.com/DUNE/lblpwgtools/tree/feature/ndlar-cafs-wip) that disables it that can be used until this can be resolved.

### CODE REVIEW

File: src/testHDF.C [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    std::cout << "Tracks for event " << evtIdx << ":" << std::endl;
+    for (const auto & trk : sr.ndlar.tracks)
+    {
+      std::cout << "  start = " << trk.start
+                << ",  end = " << trk.end
+                << ",  end dir = " << trk.end_dir
+                << std::endl;
+    }
+  }
+}
```
Comment (cjbackhouse) [inline src/testHDF.C:43]: Can we appease the various tools that want files to end in a newline?

File: src/CAF.cxx [hunk 2]
```diff
@@ -0,0 +1,90 @@
+#define CAF_cxx
+#ifdef CAF_cxx
```
Comment (cjbackhouse) [inline src/CAF.cxx:2]: Why include guards in a cxx file?

Comment (chenel) [reply src/CAF.cxx:2]: ha.  this must have been inherited from some previous edition of the file...  I think it was previously being `#include`d into a macro or something, but obviously not needed any more now that I've straightened that out.

Comment (chenel) [reply src/CAF.cxx]: taken care of in 20d6778c

File: src/testHDF.C (hunk 1 above)
Comment (chenel) [reply src/testHDF.C]: came as a drive-by in commit 6a209d2a

File: build.sh [hunk 3]
```diff
@@ -17,6 +17,8 @@ setup genie_phyopt v2_12_10   -q dkcharmtau
 setup geant4 v4_10_3_p01b -q e15:prof
 setup jobsub_client
 setup eigen v3_3_5
+setup duneanaobj v01_01_00 -q e15:gv1:debug
```
Comment (cjbackhouse) [inline build.sh:20]: What is the motivation for `debug` here? I think you would want all one or all the other.

Comment (chenel) [reply build.sh]: yeah, this build file is kind of a mess actually (you'll notice `dk2nugenie` and `geant4` have different quals here too).  I had been using `debug` while testing and forgot about it.  when this thing goes into production and we update to `e20` somebody should take a stab at rationalizing it more, but at least for now I've changed everything to `prof` (5f6153cb).

File: Makefile [hunk 4]
```diff
@@ -52,17 +59,28 @@ LDLIBS += -L$(GENIE)/lib \
...
-	rm -f $(wildcard *.o) $(patsubst %.cxx, %, $(wildcard *.cxx))
-	rm -f $(wildcard AutoDict_*)
\ No newline at end of file
+	rm -f $(LIBDIR)/*
+	rm -f $(BINDIR)/*
+	rm -f $(wildcard AutoDict_*)
+	+make -C src clean
+
+test:
+	+make -C src $(MAKECMDGOALS)
```
[the remark this answers is not in the document]
Comment (chenel) [reply Makefile]: yeah, yeah, ok.  92df5c5

**Review by cjbackhouse — APPROVED** [review]
Most of the actual caf-making details are over my head, but I'm happy with the broad outline. It seems more likely someone else should look at this than it did for duneanaobj, but I'm not sure if it's actually necessary.

**Comment by chenel** [comment]
> Most of the actual caf-making details are over my head, but I'm happy with the broad outline. It seems more likely someone else should look at this than it did for duneanaobj, but I'm not sure if it's actually necessary.

I believe @muether and @lcrem are responsible for figuring out who should officially review here.

File: src/truth/FillTruth.cxx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  // Add DUNErw weights to the CAF
+  sr.total_xsSyst_cv_wgt = 1;
+  // fixme: the following is disabled until DIRT-II finishes on model + uncertainty decisions
+  //systtools::event_unit_response_w_cv_t resp = rh.GetEventVariationAndCVResponse(*event);
+  //for( const systtools::VarAndCVResponse& it : resp ) {
+  //  // Need begin/end to convert double to float
+  //  sr.xsSyst_wgt.emplace_back(it.responses.begin(), it.responses.end());
+  //  sr.cvwgt.push_back(it.CV_response);
+  //  sr.total_xsSyst_cv_wgt *= it.CV_response;
+  //}
```
Comment (chenel) [inline src/truth/FillTruth.cxx:133]: @luketpickering these are the nusystematics lines in question

**Comment by chenel** [comment]
As per discussion at the May 2022 collaboration meeting, I'm merging this now.  The previous version is tagged as [`last_unstructured_cafs`](https://github.com/DUNE/ND_CAFMaker/releases/tag/last_unstructured_cafs).

**Note that the `nusystematics` plugins are disabled for now until an expert can look at them**

**Pull request merged**